source/extensions/filters/network/rocketmq_proxy/app_load_balancer.h

```diff
@@ -0,0 +1,59 @@
+#pragma once
+
+#include <string>
+
+#include "envoy/common/pure.h"
+
+#include "common/common/logger.h"
+
+#include "extensions/filters/network/rocketmq_proxy/protocol.h"
+
+namespace Envoy {
+namespace Extensions {
+namespace NetworkFilters {
+namespace RocketmqProxy {
+
+class MessageQueue;
+
+/**
+ * Application load balancer handles load balancing requirement in RocketMQ protocol tier.
+ * The reason why this feature is not implemented using envoy LB facility has two parts. The first
+ * one is concept mismatch between RocketMQ and envoy. The second one is this proxy implement makes
+ * most of the Delta CDS.
+ */
+class AppLoadBalancer : public Logger::Loggable<Logger::Id::rocketmq> {
```

Comment:
Can you tell me specifically why envoy LB can't be used instead?

Reply:
In envoy, hosts within a cluster are functionally identical. Load balancer algorithms are all about hosts. In RocketMQ, things are a little bit subtle. We have topic-host-queue 3 tiers involved when performing load balancer. Aka, somethings, we need to take queue into consideration when performing canary consumption.

Reply:
I don't quite agree with customizing my own load balancer here, and I prefer to extend it on Envoy's load balancer

Reply:
@zyfjeff  OK, I would remove this part.